lib/segment/src/index/hnsw_index/graph_links.rs

```diff
@@ -140,81 +304,289 @@ impl GraphLinks {
-        let end = self.offsets[layer_offsets_start + reindexed_point_id + 1];
-        &self.links[start..end]
-    }
-
     fn get_level_offsets_range(&self, level: usize) -> Option<Range<usize>> {
-        if level < self.level_offsets.len() {
-            let layer_offsets_start = self.level_offsets[level];
-            let layer_offsets_end = if level + 1 < self.level_offsets.len() {
+        if level < self.levels_count() {
+            let layer_offsets_start = self.get_level_offset(level);
+            let layer_offsets_end = if level + 1 < self.levels_count() {
                 // `level` is not last, next level_offsets is end of range
-                self.level_offsets[level + 1]
+                self.get_level_offset(level + 1)
             } else {
                 // `level` is last, next `offsets.len()` is end of range
-                self.offsets.len() - 1
+                self.offsets_len() - 1
             };
             Some(layer_offsets_start..layer_offsets_end)
         } else {
             None
         }
     }
+}
 
-    fn fill_level_links<I>(
-        &mut self,
-        level: usize,
-        level_points_iter: I,
-        edges: &[Vec<Vec<PointOffsetType>>],
-    ) where
-        I: Iterator<Item = usize>,
-    {
-        self.level_offsets.push(self.offsets.len() - 1);
+#[derive(Default)]
+pub struct GraphLinksRam {
+    // all flattened links of all levels
+    links: Vec<PointOffsetType>,
+    // all ranges in `links`. each range is `links[offsets[i]..offsets[i+1]]`
+    // ranges are sorted by level
+    offsets: Vec<u64>,
+    // start offet of each level in `offsets`
+    level_offsets: Vec<u64>,
+    // for level 1 and above: reindex[point_id] = index of point_id in offsets
+    reindex: Vec<PointOffsetType>,
+}
+
+impl GraphLinks for GraphLinksRam {
+    fn load_from_file(path: &Path) -> OperationResult<Self> {
+        let mmap = GraphLinksMmap::load_from_file(path)?;
+
+        Ok(Self {
+            links: mmap.get_links_slice().to_vec(),
+            offsets: mmap.get_offsets_slice().to_vec(),
+            level_offsets: mmap.level_offsets.clone(),
+            reindex: mmap.get_reindex_slice().to_vec(),
+        })
+    }
+
+    fn from_vec(
+        edges: Vec<Vec<Vec<PointOffsetType>>>,
+        path: Option<&Path>,
+    ) -> OperationResult<Self> {
+        let mut graph_links = GraphLinksRam {
+            links: Vec::new(),
+            offsets: vec![0],
+            level_offsets: Vec::new(),
+            reindex: Vec::new(),
+        };
 
-        for point_id in level_points_iter {
-            let links = &edges[point_id][level];
-            self.links.extend_from_slice(links);
-            self.offsets.push(self.links.len());
+        let mut converter = GraphLinksConverter::new(edges);
+        if let Some(path) = path {
+            converter.save_to_file(path)?;
+        }
+
+        if !converter.edges.is_empty() {
```

Comment:
This code of converting `edges` into `graph_links` is only used once, during the initial conversion. It is also the one used in most tests, so we don't actually test the code which is used most in real-life scenario. The proposition is to avoid duplication and use same cove for conversion in all cases